server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocatorTests.java

```diff
@@ -333,6 +336,73 @@ public ShardAllocationDecision decideShardAllocation(ShardRouting shard, Routing
         assertThat(listenersCalled.get(), equalTo(1));
     }
 
+    public void testConcurrency() throws Exception {
+
+        var createdIndices = new CopyOnWriteArraySet<String>();
+        var allocatedIndices = new CopyOnWriteArraySet<String>();
+        var reroutedIndices = new CopyOnWriteArraySet<String>();
+
+        var threadPool = new TestThreadPool(getTestName());
+        RerouteService rerouteService = (r, p, l) -> threadPool.executor(ThreadPool.Names.CLUSTER_COORDINATION).submit(() -> {
+            reroutedIndices.addAll(createdIndices);
+            l.onResponse(null);
+        });
+        var allocator = new ShardsAllocator() {
+            @Override
+            public void allocate(RoutingAllocation allocation) {// TODO attempt to count up-to-date allocations
+                final var dataNodeId = allocation.nodes().getDataNodes().values().iterator().next().getId();
+                final var unassignedIterator = allocation.routingNodes().unassigned().iterator();
+                while (unassignedIterator.hasNext()) {
+                    String indexName = unassignedIterator.next().getIndexName();
+                    allocatedIndices.add(indexName);
+                    unassignedIterator.initialize(dataNodeId, null, 0L, allocation.changes());
+                }
+            }
+
+            @Override
+            public ShardAllocationDecision decideShardAllocation(ShardRouting shard, RoutingAllocation allocation) {
+                throw new AssertionError("only used for allocation explain");
+            }
+        };
+        var desiredBalanceShardsAllocator = new DesiredBalanceShardsAllocator(allocator, threadPool, () -> rerouteService);
+
+        var discoveryNode = createDiscoveryNode();
+        var clusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .nodes(DiscoveryNodes.builder().add(discoveryNode).localNodeId(discoveryNode.getId()).masterNodeId(discoveryNode.getId()))
+            .build();
+
+        var indexNameGenerator = new AtomicInteger();
+        var listenersCalled = new AtomicInteger();
+
+        var iterations = 1_000;
```

Comment:
Think we can randomise the number of iterations, and also use a `CountDownLatch` to avoid the `assertBusy()` at the end?

```suggestion
        var iterations = between(1, 1000);
        var listenersCountdown = new CountDownLatch(iterations);
```